tools/network/dnssec/trustchain.go

```diff
@@ -0,0 +1,364 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package dnssec
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/algorand/go-deadlock"
+	"github.com/miekg/dns"
+)
+
+type trustChain struct {
+	resolver     netResolverIf
+	trustedZones map[string]*trustedZone
+	mu           deadlock.RWMutex
+}
+
+type trustedZone struct {
+	name  string
+	zsk   map[uint16]dns.DNSKEY // zone signing keys
+	ksk   map[uint16]dns.DNSKEY // key signing keys
+	rrSig map[uint16]dns.RRSIG  // DNSKEY RR signature(s), used for validity
+	dss   []dns.DS              // DS or root anchor authenticating this zone in DS format
+}
+
+func makeTrustChain(r netResolverIf) *trustChain {
+	return &trustChain{
+		resolver:     r,
+		trustedZones: make(map[string]*trustedZone),
+	}
+}
+
+// verifyRRSig takes RRSET, array of RRSIG, time and ZSKs in form of map
+// and returns a map of matched signatures by keytag
+func verifyRRSig(rrSet []dns.RR, rrSig []dns.RRSIG, t time.Time, keys *map[uint16]dns.DNSKEY) map[uint16]dns.RRSIG {
+	verifiedSig := make(map[uint16]dns.RRSIG)
+	for _, sig := range rrSig {
+		if !sig.ValidityPeriod(t) {
+			continue
+		}
+		ksk := (*keys)[sig.KeyTag]
+		if err := sig.Verify(&ksk, rrSet); err == nil {
+			verifiedSig[sig.KeyTag] = sig
+		}
+	}
+	return verifiedSig
+}
+
+// verifyKSKDigest takes array of DS and KSKs in form of map
+// and returns an array of matched DS records and verified keys
+func verifyKSKDigest(dss *[]dns.DS, ksk *map[uint16]dns.DNSKEY) ([]dns.DS, map[uint16]dns.DNSKEY) {
```

Comment:
I'm wondering if the signature of this method could be simplified from
`func verifyKSKDigest(dss *[]dns.DS, ksk *map[uint16]dns.DNSKEY) `
to
`func verifyKSKDigest(dss []dns.DS, ksk map[uint16]dns.DNSKEY) ` ?

I don't see any obvious issue, but as before, it's uncommon to use pointers to slices or maps.